tools/litex-ci-runner/src/main.rs

```diff
@@ -0,0 +1,264 @@
+    // application.
+    tmpl_run(&["tests/mpu_walk_region"], |p| {
+        p.exp_string("[TEST] MPU Walk Regions")?;
+
+        // Start walking the entire flash region, should not panic
+        p.exp_string("Walking flash")?;
+
+        // If we receive this string, walking the flash region
+        // worked. Start walking the entire memory region, also should
+        // not panic.
+        p.exp_string("Walking memory")?;
+        println!("Walking flash worked.");
+
+        // One final time, expect "Walking flash", which indicates
+        // that walking the memory region worked.
+        p.exp_string("Walking flash")?;
+        println!("Walking memory worked.");
+
+        Ok(())
+    })?;
+
+    // TODO: simulate a button press to overrun both the memory and
+    // flash sections to verify that the MPU protections indeed work
+
+    // Test: c_hello and printf_long
+    //
+    // Most importantly, tests whether multiple apps can run on the
+    // board. Furthermore tests whether the console (TX) is properly
+    // muxed between applications.
+    tmpl_run(&["c_hello", "tests/printf_long"], |p| {
+        // Output may arrive out of order, but verify that all three
+        // messages arrive fully, and the second message of
+        // printf_long arrives after the first one.
+        let (_, matched) = p.exp_regex("(Hi welcome to Tock\\. |Hello World!)")?;
+        if matched == "Hi welcome to Tock. " {
+            let (_, matched) = p.exp_regex("(Hello World!|This test makes sure )")?;
+            if matched == "Hello World!" {
+                p.exp_string(
+                    "This test makes sure that a greater than 64 byte message can be printed.",
+                )?;
+            } else {
+                p.exp_string("that a greater than 64 byte message can be printed.")?;
+            }
+        } else {
+            p.exp_string("Hi welcome to Tock. This test makes sure that a greater than 64 byte message can be printed.")?;
+            p.exp_string("And a short message.")?;
+        }
+
+        Ok(())
+    })?;
+
+    // Test: rot13_client and rot13_service
+    //
+    // Tests IPC and IPC service discovery.
+    tmpl_run(&["rot13_service", "rot13_client"], |p| {
+        // Let run for a few cycles to make sure it doesn't crash
+        // after the first few messages
+        for _ in 0..10 {
+            p.exp_string("12: Uryyb Jbeyq!")?;
+            p.exp_string("12: Hello World!")?;
+        }
+
+        Ok(())
+    })?;
+
+    // Test: console_recv_short and console_recv_long
+    //
+    // This is expected to fail. The console RX is not virtualized
+    // yet, and thus one of the apps trying to receive over the
+    // console will get an error from the kernel.
+    tmpl_run(
+        &["tests/console_recv_short", "tests/console_recv_long"],
+        |p| {
+            p.exp_regex("\\[(SHORT|LONG)\\] Error doing UART receive: -2")?;
+            Ok(())
+        },
+    )?;
```

Comment:
I don't think we should include a test that would cause a PR that actually fixes this problem to fail.

Reply:
I included this so it would fail and we can think of a proper test mechanism for the then fixed version. So it's essentially a reminder that we should write tests for it.

Reply:
I guess I don't understand but if you think this is important to track then 👍